From 696f7c4812544dd9a0775a354768abca715dee79 Mon Sep 17 00:00:00 2001
From: Shilin Victor <chrono.monochrome@gmail.com>
Date: Sun, 17 Sep 2017 12:30:52 +0300
Subject: [PATCH 2/7] hwui: Allow devices to opt-out of asynchronouse
 setSurface initialization

Not all devices seem to be able to deal with this so let them opt
out of it, even though it does provide a UI slow-down.

Change-Id: I0d94c333c3c4bdf401261455012e422fa0d185a2
---
 libs/hwui/Android.mk                   | 4 ++++
 libs/hwui/renderthread/RenderProxy.cpp | 4 ++++
 2 files changed, 8 insertions(+)

diff --git a/libs/hwui/Android.mk b/libs/hwui/Android.mk
index c54feeb2..59a0acbb 100644
--- a/libs/hwui/Android.mk
+++ b/libs/hwui/Android.mk
@@ -118,6 +118,10 @@ ifeq ($(TARGET_USES_HWC2),true)
     hwui_cflags += -DUSE_HWC2
 endif
 
+ifeq ($(TARGET_REQUIRES_SYNCHRONOUS_SETSURFACE),true)
+    hwui_cflags += -DREQUIRES_SYNCHRONOUS_SETSURFACE
+endif
+
 # GCC false-positives on this warning, and since we -Werror that's
 # a problem
 hwui_cflags += -Wno-free-nonheap-object
diff --git a/libs/hwui/renderthread/RenderProxy.cpp b/libs/hwui/renderthread/RenderProxy.cpp
index ce13ebc5..14f2feab 100644
--- a/libs/hwui/renderthread/RenderProxy.cpp
+++ b/libs/hwui/renderthread/RenderProxy.cpp
@@ -146,7 +146,11 @@ void RenderProxy::initialize(const sp<Surface>& surface) {
     SETUP_TASK(initialize);
     args->context = mContext;
     args->surface = surface.get();
+#ifdef REQUIRES_SYNCHRONOUS_SETSURFACE
+    postAndWait(task);
+#else
     post(task);
+#endif
 }
 
 CREATE_BRIDGE2(updateSurface, CanvasContext* context, Surface* surface) {
-- 
2.17.1

